### DISCUSSION TITLE
Replacement of `build_library`

### DISCUSSION THREAD

**Discussion by balatic** [tree-sitter/py-tree-sitter#241] [Q&A] [answered]
@ObserverOfTime  First of all, amazing work. When upgrading from `0.21.3` to `0.22.2`, it seems like a lot has changed in the way we build grammar files, and in the way we access the parser. What replaced `Language.build_library`? And what's the source of the so called ready to use pip packages like `tree_sitter_python`? Do these pip packages have any open source repo? Or are they being published privately? How do we load grammar files other than the privately sourced ones produced by `amaanq`?

**Comment by ObserverOfTime** [comment]
You can find them on [pypi](https://pypi.org/search/?q=tree-sitter&o=&c=Topic+%3A%3A+Software+Development+%3A%3A+Compilers&c=Topic+%3A%3A+Text+Processing+%3A%3A+Linguistic).

**Reply by balatic** [reply]
I can see the wheels there, but not the repos where they are built from. Wanted to look at their source, to build our grammar files.

**Reply by WesleyYue** [reply]
@ObserverOfTime given that a lot of community maintained parsers are not yet on pypi, it feels premature to remove build_library. We're building a general coding copilot which means we should have a good sample of languages being used in the wild. For us, over half of the languages we support are not yet on pypi and this prevents us from moving to 0.22

**Reply by jhandley** [reply]
I agree. We use Kotlin and Swift, neither of which are there.

**Reply by ObserverOfTime** [reply]
You can stay on 0.21 or you can ask/help the parsers to update.

**Reply by issacl2c** [reply]
tree-sitter inherently allows the parser grammar to be updated and used. Please bring back this feature.

**Comment by TheTeaCat** [comment]
@ObserverOfTime Why was `build_library` removed?

**Comment by WesleyYue** [comment, marked as answer]
@balatic to answer your question I believe the pypi wheels are built from github actions: https://github.com/tree-sitter/py-tree-sitter/blob/master/.github/workflows/pypi.yml

**Comment by issacl2c** [comment]
@ObserverOfTime can you please suggest a way for us to update the grammar at runtime?
Our usecases are for custom macros, and these  custom macros need to be updated to the grammer on the fly.

**Reply by ObserverOfTime** [reply]
No. That was never supported but maybe @amaanq has an idea.

**Comment by Avinier** [comment]
Hi @WesleyYue what's the progress on the general copilot? Also what is the best implementation procedure to replace `build_library` ?

**Comment by Trevoke** [comment]
I am trying to build a grammar that doesn't yet exist, and I would like to do so in Python. I am quite saddened that the python bindings won't allow me to do that going forward.

The answer "downgrade to 0.21" should probably be the accepted answer in this discussion.
